File: cpp/Graph/133.clone-graph.cpp
#include "../std.hpp"
class Solution {
public:
    unordered_map<Node *, Node *> m;
    Node *cloneGraph(Node *node)
    {
        if (!node) {
            return nullptr;
        }
        if (m.find(node) == m.end()) {
            m[node] = new Node(node->val);
            for (auto &&n : node->neighbors) {
                m[node]->neighbors.push_back(cloneGraph(n));
            }
        }
        return m[node];
    }
};

File: TypeScript/Graph/133.clone-graph.ts
class Node {
  val: number
  neighbors: Node[]
  constructor(val?: number, neighbors?: Node[]) {
    this.val = val === undefined ? 0 : val
    this.neighbors = neighbors === undefined ? [] : neighbors
  }
}
function cloneGraph(node: Node | null): Node | null {
  let visited = {}
  function dfs(node: Node | null) {
    if (!node) {
      return node
    }
    if (visited[node.val]) {
      return visited[node.val]
    }
    const copy = new Node(node.val)
    visited[node.val] = copy
    node.neighbors.forEach(n => {
      copy.neighbors.push(dfs(n))
    })
    return copy
  }
  return dfs(node)
}
